### PR TITLE
Store angles internally as radians

### PR DESCRIPTION
**Pull request by vittorioromeo**
Discuss in #2976

### CODE REVIEW

**Comment by Bambo-Borris** [comment]
Personally, I can't see a downside to this change. Think it makes sense, and is a nice, small improvement atop the already great work done by adding `sf::Angle` to the API.

File: include/SFML/System/Angle.hpp [hunk 1]
```diff
@@ -507,4 +519,9 @@ namespace Literals
 /// sf::Angle a6 = 3.14_rad; // 3.14 radians
 /// \endcode
 ///
+/// Internally, sf::Angle stores the angle value as radians.
+///
+/// For the best performance and to minimize run-time conversions,
+/// it is recommended to work with radians whenever possible.
+///
```
Comment (ChrisThrasher) [inline include/SFML/System/Angle.hpp:526]: I don't necessarily recommend this. I want users to do whatever they find most ergonomic. Besides, it breaks encapsulation to start making recommendations based on implementation details.

File: src/SFML/Audio/SoundSource.cpp [hunk 2]
```diff
@@ -25,6 +25,7 @@
 ////////////////////////////////////////////////////////////
 // Headers
 ////////////////////////////////////////////////////////////
+#include "SFML/System/Angle.hpp"
```
Comment (ChrisThrasher) [inline src/SFML/Audio/SoundSource.cpp:28]:
```suggestion
```

`SoundSource::Cone` already uses `sf::Angle` to define its data layout so we can guarantee it's included in SoundSource.hpp. I'm guessing something like clangd automatically added this as evidence by the quotes instead of angle brackets.

File: test/Graphics/Transformable.test.cpp [hunk 3]
```diff
@@ -34,11 +34,11 @@ TEST_CASE("[Graphics] sf::Transformable")
...
 
         transformable.setRotation(sf::degrees(3.14f));
-        CHECK(transformable.getRotation() == sf::degrees(3.14f));
+        CHECK(transformable.getRotation() == Approx(sf::degrees(3.14f)));
         transformable.setRotation(sf::degrees(540));
-        CHECK(transformable.getRotation() == sf::degrees(180));
+        CHECK(transformable.getRotation() == Approx(sf::degrees(180)));
         transformable.setRotation(sf::degrees(-72));
-        CHECK(transformable.getRotation() == sf::degrees(288));
+        CHECK(transformable.getRotation() == Approx(sf::degrees(288)));
```
Comment (ChrisThrasher) [inline test/Graphics/Transformable.test.cpp:41]: Let's change the `setRotation` arguments to use `sf::radians` so we can avoid adding approximate comparisons.

File: test/System/Angle.test.cpp [hunk 4]
```diff
@@ -202,27 +202,27 @@ TEST_CASE("[System] sf::Angle")
...
-            STATIC_CHECK(sf::degrees(10) * 2.5f == sf::degrees(25));
-            STATIC_CHECK(sf::degrees(100) * 10.0f == sf::degrees(1000));
+            CHECK(sf::degrees(10) * 2.5f == Approx(sf::degrees(25)));
+            CHECK(sf::degrees(100) * 10.0f == Approx(sf::degrees(1000)));
 
             STATIC_CHECK(10 * sf::radians(0) == sf::Angle::Zero);
-            STATIC_CHECK(2.5f * sf::degrees(10) == sf::degrees(25));
-            STATIC_CHECK(10.0f * sf::degrees(100) == sf::degrees(1000));
+            CHECK(2.5f * sf::degrees(10) == Approx(sf::degrees(25)));
+            CHECK(10.0f * sf::degrees(100) == Approx(sf::degrees(1000)));
```
Comment (ChrisThrasher) [inline test/System/Angle.test.cpp:218]: Anywhere we previously did exact comparison with degrees I'd rather we do exact comparisons with radians rather than start doing approximate comparisons.

File: test/System/Angle.test.cpp [hunk 5]
```diff
@@ -25,33 +25,33 @@ TEST_CASE("[System] sf::Angle")
...
+        CHECK(sf::degrees(1).wrapUnsigned() == Approx(sf::degrees(1)));
+        CHECK(sf::degrees(-1).wrapUnsigned() == Approx(sf::degrees(359)));
+        CHECK(sf::degrees(90).wrapUnsigned() == Approx(sf::degrees(90)));
+        CHECK(sf::degrees(-90).wrapUnsigned() == Approx(sf::degrees(270)));
+        CHECK(sf::degrees(180).wrapUnsigned() == Approx(sf::degrees(180)));
+        CHECK(sf::degrees(-180).wrapUnsigned() == Approx(sf::degrees(180)));
+        CHECK(sf::degrees(360).wrapUnsigned() == Approx(sf::degrees(0)));
+        CHECK(sf::degrees(-360).wrapUnsigned() == Approx(sf::degrees(0)));
+        CHECK(sf::degrees(720).wrapUnsigned() == Approx(sf::degrees(0)));
+        CHECK(sf::degrees(-720).wrapUnsigned() == Approx(sf::degrees(0)));
```
Comment (ChrisThrasher) [inline test/System/Angle.test.cpp:54]: These changes show how the new API is perhaps less intuitive. Humans have an easier time reasoning about angles in degrees. It's not intuitive to me that 360 degrees wrapped doesn't yield precisely 0 degrees. At least with radians we know that pi cannot be perfectly represented by floating point numbers so it's not confusing why some number close to 2pi doesn't wrap exactly to 0 radians.

**Review by ChrisThrasher — COMMENTED** [review]
I'm not sure whether I prefer this new implementation. While this change may theoretically lead to fewer internal conversions, it also means using this API in terms of degrees yields results that are less exact than before.

File: test/System/Angle.test.cpp (hunk 4 above)
Comment (L0laapk3) [reply test/System/Angle.test.cpp:218]: Just wanted to put in my two cents because I disagree with this point.

In my opinion, all of these tests should do approx comparisons either way as with any other float test.

By testing for exact values of float, you are no longer only testing if the API works but also testing for implementation details.

File: test/System/Angle.test.cpp (hunk 5 above)
Comment (L0laapk3) [reply test/System/Angle.test.cpp:54]: float additions never guaranteed no rounding errors and you will never be able to guarantee that 360 degrees wrapped is exactly 0 degrees. It just happened to be that way for these tests but the result was technically speaking undefined behavior.

floating point rounding errors are a very basic concept in computer programming and I feel like if we need to protect users from them in arbitrary situations then we are doing it wrong.

File: include/SFML/System/Angle.hpp (hunk 1 above)
Comment (L0laapk3) [reply include/SFML/System/Angle.hpp:526]: and also, in many cases there is no cost as the compiler will convert constants at compile time.

File: test/System/Angle.test.cpp (hunk 5 above)
Comment (ChrisThrasher) [reply test/System/Angle.test.cpp:54]: > but the result was technically speaking undefined behavior.

Its we're going to get into the pedantry of this, it is not undefined. IEEE float math is perfectly well defined.

> you will never be able to guarantee that 360 degrees wrapped is exactly 0 degrees.

Actually yes we can guarantee that `360.f` will wrap to `0.f`. Floats do a perfect job representing integers. It's once you start using values in between integers or try to use really big numbers that the "weird" stuff comes into play.

Comment (L0laapk3) [reply test/System/Angle.test.cpp:54]: You are indeed correct that floating point arithmetic rounding errors are well defined, i was not aware of this.

Still, the whole concept of relying on float arithmetic to give an exact result is wild to me.

It is not defined in the sf::Angle documentation anywhere that it will give exact results for some arithmetic operations, therefore in my opinion by not using approx comparisons breaks encapsulation by testing implementation details.

**Comment by Bambo-Borris** [comment]
> I'm not sure whether I prefer this new implementation. While this change may theoretically lead to fewer internal conversions, it also means using this API in terms of degrees yields results that are less exact than before.

This may very well be true, but given the usecases of SFML within games is quite prevalent, the use of radians is preferred in this scenario in my view. Also tying back to the fact the standard library maths functions also prefer radians, SFML internally using radians seems sensible. The fact it will be slightly less exact when you do `sf::degrees(90.f)` is really going to be a question of what scale does it take for a user to notice this inexact angle? I think by the time it becomes noticeably inexact, there are probably other visual problems a user has that far exceed the angle problem. 

If we look to raylib it requires users pass an angle in degrees, but internally always convert that to radians for use in their own maths libraries. GLM will always require radians for rotations. I think the fact we have this nice abstraction means users can still work within the realm of degrees, but internally I think it's sensible for SFML to utilise radians.

**Comment by L0laapk3** [comment]
> Other than the iOS sensor impl I'm not really seeing any less conversions, so does this actually change anything?
> 
> As far as I can tell the only motivation is that some people may think the implementation is unintuitive, but the same could easily be said for using radians. It seems subjective without any technical reason to change

I have linked a few now removed conversions here: https://github.com/SFML/SFML/issues/2976

**Comment by vittorioromeo** [comment]
> Other than the iOS sensor impl I'm not really seeing any less conversions, so does this actually change anything?

Every call to `sf::Angle::asRadians` used to incur a conversion cost prior to this PR.

**Comment by JonnyPtn** [comment]
And every call to `asDegrees` will in cur a conversion cost with this PR too right? So it's an arbitrary decision really

**Comment by vittorioromeo** [comment]
> And every call to `asDegrees` will in cur a conversion cost with this PR too right? So it's an arbitrary decision really

It's really not because most of the programming/gamedev world has already "agreed" on using radians.

**Comment by JonnyPtn** [comment]
I don't really see what relation that has, as SFML doesn't use radians or degrees in it's API now (would have been relevant with 2.x API where degrees was used). This is just about which is used internally

Worth noting that everybody upgrading to SFML 3 will be passing degrees in too due to the existing API

**Comment by eXpl0it3r** [comment]
I believe one of the main benefits is, that for certain internal calculations (anything using `<cmath>` functions), we wouldn't need to convert from degrees to radians. Meaning, every `asRadians()` that is there currently, won't require a conversion now - this doesn't necessarily show up in the PR, as no change is needed.

As for the API usage perspective, I don't think there's a point in making guesses in who and how many are using which API. The point of `sf::Angles` is people are free to choose, thus there will always be one or the other that requires a conversion.

File: include/SFML/System/Angle.hpp [hunk 6]
```diff
@@ -138,26 +138,32 @@ class Angle
     ////////////////////////////////////////////////////////////
     // NOLINTNEXTLINE(readability-identifier-naming)
     static const Angle Zero; //!< Predefined 0 degree angle value
+    // NOLINTNEXTLINE(readability-identifier-naming)
+    static const Angle QuarterCircle; //!< Predefined 90 degree angle value
+    // NOLINTNEXTLINE(readability-identifier-naming)
+    static const Angle HalfCircle; //!< Predefined 180 degree angle value
+    // NOLINTNEXTLINE(readability-identifier-naming)
+    static const Angle FullCircle; //!< Predefined 360 degree angle value
```
Comment (eXpl0it3r) [inline include/SFML/System/Angle.hpp:146]: Please always highlight new APIs in the PR description, at best keep them as separate commit or even PR.

**Comment by vittorioromeo** [comment]
> using `<cmath>` functions), we wouldn't need to convert from degrees to radians. Meaning, every `asRadians()` that is there currently, won't require a conversion now - this doesn't necessarily show up in the PR, as no change is needed.
> 
> As for the API usage perspective, I don't think there's a point in making

SFML can be and is often used in conjuction with other libraries such as Vulkan, glm, or even the Standard Library that all expect angles in radians.

File: include/SFML/System/Angle.hpp (hunk 6 above)
Comment (ChrisThrasher) [reply include/SFML/System/Angle.hpp:146]: I'd also rather save such additions for a later PR as to keep this PR as minimal as possible

Comment (vittorioromeo) [reply include/SFML/System/Angle.hpp:146]: OK, will remove for a later PR, I think these are beneficial to avoid using `degrees` and relying on compile-time conversions that might not be done at compile-time

File: include/SFML/System/Angle.hpp (hunk 1 above)
Comment (vittorioromeo) [reply include/SFML/System/Angle.hpp:526]: I think it's good to have this but will leave for a future PR to minimize friction in this one

File: test/Graphics/Transformable.test.cpp (hunk 3 above)
Comment (vittorioromeo) [reply test/Graphics/Transformable.test.cpp:41]: I gave it a try and some tests still failed due to small floating point precision errors. I don't think it's worthwhile spending time on avoiding the use of `Approx` in the test, I agree with @L0laapk3's technical arguments that `Approx` is what we want to test for, not an exact value.

**Comment by vittorioromeo** [comment]
Removed the controversial changes

**Comment by ChrisThrasher** [comment]
Did a quick accounting of where we need angles as radians versus where we need angles as degrees to help understand which format gets more use, at least within SFML itself. Is it true we use `.asRadians()` to a similar degree as `.asDegrees()`? Let's read the code.

As radians:
- Trigonometry in 4 example programs (see https://github.com/SFML/SFML/pull/2028 for a better API for common trig with `sf::Angle`)
- Use with miniaudio which expects radians
- Italic text calculations in `sf::Text`
- 2D matrix math in `sf::Transform`
- 2D matrix math in `sf::Transformable`
- 2D matrix math in `sf::View`
- 2D matrix math in `sf::Vector2<T>`

As degrees:
- Reporting gyroscope rates on iOS as degrees per second (See #3001 for why this may actually be a bug)

As it stands now users are getting free construction from degrees and then repeated conversions to radians under the hood since as evident by the source code. Upon changing to a radians-based implementation, construction from degrees is possibly the only place where a conversion between degrees and radians (in either direction) would be necessary. We have to expect users will continue to prefer writing `sf::degrees(45)` rather than `sf::radians(0.78539816339)` since humans have a much easier time reasoning about angles in terms of degrees. Not necessarily saying anything new but I wanted to provide some more evidence for this fact. There is seemingly no place in the code (outside of that one iOS snippet) where a degree-based implementation reduces the number of degree<->radian conversions.

The only place degrees would be superior is if a user has lots of code that performs arithmetic of angles in terms of degrees. This is code that theoretically exists but I have no evidence to believe that users are in fact routinely writing code like `sf::degrees(360) % sf::degrees(15) + sf::degrees(45)` or anything like that so I can't lean too heavily on previous arguments I made to this end.

I'll also mention as a historical fact that `sf::Angle` is currently implemented in terms of degrees because SFML 2 represented all angles as degrees. We did not ever discuss changing the underlying representation of angles in that PR. It was not something that ever crossed our mind if I recall correctly.

---

In conclusion, my hesitation about a radians-based implementation are probably outweighed by the benefits of performing fewer casts between radians and degrees. I think I'm leaning in favor of this PR, certainly enough so that I don't want to block it. If the team is in favor then let's go for it.

**Comment by ChrisThrasher** [comment]
SFMLBuildMaster: Build this please.

File: src/SFML/Audio/AudioDevice.hpp [hunk 7]
```diff
@@ -27,6 +27,8 @@
 ////////////////////////////////////////////////////////////
 // Headers
 ////////////////////////////////////////////////////////////
+#include "SFML/System/Angle.hpp"
```
Comment (ChrisThrasher) [inline src/SFML/Audio/AudioDevice.hpp:30]: Looks like another automatic inclusion from your IDE or language server. We don't need this because Listener.hpp already includes Angle.hpp.

**Comment by kimci86** [comment]
> We have to expect users will continue to prefer writing sf::degrees(45) rather than sf::radians(0.78539816339) since humans have a much easier time reasoning about angles in terms of degrees.

It is worth noting that the conversion can happen at compile time when using hardcoded contant angles or constant expressions because Angle uses constexpr everywhere.

**Review by ChrisThrasher — CHANGES_REQUESTED** [review]
The core of the PR is good.

My one nit is your use of `.f` suffixes. We don't need to do that. `sf::degrees(360)` is perfectly safe and well defined. No need to add `.f` in so many places. I find it no easier to read and it has no technical benefits.

File: src/SFML/Audio/AudioDevice.hpp (hunk 7 above)
Comment (vittorioromeo) [reply src/SFML/Audio/AudioDevice.hpp:30]: Sorry, removing

**Review by ChrisThrasher — APPROVED** [review]

File: include/SFML/System/Angle.inl [hunk 8]
```diff
@@ -34,16 +34,14 @@ namespace sf
 {
 namespace priv
 {
-constexpr float pi = 3.141592654f;
+constexpr float pi  = 3.141592654f;
```
Comment (L0laapk3) [inline include/SFML/System/Angle.inl:37]: I wanted to suggest `std::numbers::pi_v` but I guess we are not C++20.

File: include/SFML/System/Angle.inl [hunk 9]
```diff
@@ -55,104 +53,104 @@ constexpr Angle::Angle() = default;
 ////////////////////////////////////////////////////////////
 constexpr float Angle::asDegrees() const
 {
-    return m_degrees;
+    return m_radians * (180.f / priv::pi);
```
Comment (L0laapk3) [inline include/SFML/System/Angle.inl:56]: Note that these constructs [already carry a baked in rounding error](https://godbolt.org/z/ThfWPfM1d).

Comment (L0laapk3) [reply include/SFML/System/Angle.inl:56]: Same with calls to `asDegrees()`, I guess it could have a double overload, I'm not sure if we have to care that much? Just pointing it out

**Review by L0laapk3 — APPROVED** [review]
just some thoughts

File: include/SFML/System/Angle.inl (hunk 9 above)
Comment (ChrisThrasher) [reply include/SFML/System/Angle.inl:56]: Are you saying that we could get marginally better accuracy if we used a single literal to represent 180 / pi rather than using this particular expression?

File: include/SFML/System/Angle.inl (hunk 8 above)
Comment (ChrisThrasher) [reply include/SFML/System/Angle.inl:37]: https://github.com/search?q=repo%3ASFML%2FSFML%20cxx_std_17&type=code

Alas still using C++17 since C++20 support isn't sufficient in the compilers we're targeting.

**Comment by eXpl0it3r** [comment]
> My one nit is your use of `.f` suffixes. We don't need to do that. `sf::degrees(360)` is perfectly safe and well defined. No need to add `.f` in so many places. I find it no easier to read and it has no technical benefits.

Personally, I like types to be explicit and not rely on implicit conversion, especially since you then have to memorize in which scenario it's not considered a narrowing conversion.

> @eXpl0it3r Any final thoughts?

I too think it's a sensible change, given there's no real reason to stick to degrees when the interface is abstracted and allows us to do fewer conversions internally.

File: include/SFML/System/Angle.inl (hunk 9 above)
Comment (eXpl0it3r) [reply include/SFML/System/Angle.inl:56]: `57.295776` - float (Compiler Explorer)  
`57.2957802` - double (Compiler Explorer)  
`57.295779513082323402053960025447` - actual (Windows calculator)

`0.000003513082323402053960025447` - difference float  
`0.000000686917676597946039974553` - difference double

I guess in theory you could also do the full math in doubles and only covert at the end, but I feel like if you care about such a precision, you may want to implement the mathy bits yourself.

The change makes sense to me, but could also be a separate PR.

**Comment by ChrisThrasher** [comment]
> > My one nit is your use of `.f` suffixes. We don't need to do that. `sf::degrees(360)` is perfectly safe and well defined. No need to add `.f` in so many places. I find it no easier to read and it has no technical benefits.
> 
> Personally, I like types to be explicit and not rely on implicit conversion, especially since you then have to memorize in which scenario it's not considered a narrowing conversion.

You don’t have to memorize anything. The compiler won’t allow narrowing conversions to happen since we have all the conversion warnings turned on.

File: include/SFML/System/Angle.hpp [hunk 10]
```diff
@@ -187,6 +187,7 @@ class Angle
 ////////////////////////////////////////////////////////////
 /// \relates Angle
 /// \brief Overload of == operator to compare two angle values
+/// \note Does not automatically wrap the angle value
```
Comment (eXpl0it3r) [inline include/SFML/System/Angle.hpp:190]: Did it do that before?

Comment (ChrisThrasher) [reply include/SFML/System/Angle.hpp:190]: This behavior did not change in this PR.

**Review by eXpl0it3r — APPROVED** [review]
